### PR TITLE
fix(napi): Make napi_wrap work on regular objects

### LINKED ISSUE #15383: Wrap finalizer for PromiseRaw failed (@napi-rs/canvas)
### What version of Bun is running?

1.1.36+ededc168c

### What platform is your computer?

Linux 5.15.153.1-microsoft-standard-WSL2 x86_64 x86_64

### What steps can reproduce the bug?

Install `@napi-rs/canvas` and make a JS file with:

```ts
import { Image } from "@napi-rs/canvas/js-binding.js"

const image = new Image();
image.src = "./anythinghere"; // doesnt need to be valid
```

### What is the expected behavior?

no error or crash, like node

### What do you see instead?

```ts
$ bun index.mjs
2 | import { Image } from "@napi-rs/canvas/js-binding.js"
3 |
4 | const image = new Image();
    ^
error: Wrap finalizer for PromiseRaw failed
 code: "ObjectExpected"

      at /index.mjs:4:1
```

### Additional information

Reported on discord: https://discord.com/channels/876711213126520882/1310131351533654016

### PR DESCRIPTION
**Pull request by 190n**
### What does this PR do?

Currently, napi_wrap only works on instances of Node-API classes. This PR will make it work on any JavaScript object.

Fixes #15383, which is occurring in `@napi-rs/canvas` and eventually in more projects because of napi-rs/napi-rs#2348

### How did you verify your code works?

Added tests and passes the existing ones.

### CODE REVIEW

**Comment by robobun** [comment]
@190n, your commit 1ff193e164fc939eac13daae34182175cb0b49d7 is building: [`#7456`](https://buildkite.com/bun/bun/builds/7456)

File: src/bun.js/bindings/napi.cpp [hunk 1]
```diff
@@ -1091,35 +1101,48 @@ extern "C" napi_status napi_remove_wrap(napi_env env, napi_value js_object,
...
-        refPtr = &val->napiRef;
-    }
+    auto* globalObject = toJS(env);
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    NapiRef* ref = nullptr;
 
-    if (!refPtr) {
-        return napi_object_expected;
+    if (!hasNapiWrap(vm, globalObject, jsc_object)) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/napi.cpp]: This calls `getDirect` twice. We only need to call it once. Let's call it once.

File: src/bun.js/bindings/napi.cpp [hunk 2]
```diff
@@ -1128,23 +1151,40 @@ extern "C" napi_status napi_unwrap(napi_env env, napi_value js_object,
...
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    const JSC::Identifier& propertyName = WebCore::builtinNames(vm).napiWrappedContentsPrivateName();
     NapiRef* ref = nullptr;
-    if (auto* val = jsDynamicCast<NapiPrototype*>(value)) {
-        ref = val->napiRef;
-    } else if (auto* val = jsDynamicCast<NapiClass*>(value)) {
-        ref = val->napiRef;
+
+    if (!hasNapiWrap(vm, globalObject, jsc_object)) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/napi.cpp]: This also calls getDirect twice

File: src/bun.js/bindings/napi.cpp [hunk 3]
```diff
@@ -1128,23 +1151,40 @@ extern "C" napi_status napi_unwrap(napi_env env, napi_value js_object,
...
+        return napi_invalid_arg;
+    }
+
+    if (napi_instance) {
+        ref = napi_instance->napiRef;
     } else {
-        ASSERT(false);
+        JSValue contents_value = jsc_object->getDirect(vm, propertyName);
+        RETURN_IF_EXCEPTION(scope, napi_pending_exception);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/napi.cpp]: We don't need to check for exceptions when using getDirect

**Pull request merged**